File: src/ban.c
/***************************************************************************
 *  Original Diku Mud copyright (C) 1990, 1991 by Sebastian Hammer,        *
 *  Michael Seifert, Hans Henrik St{rfeldt, Tom Madsen, and Katja Nyboe.   *
 *                                                                         *
 *  Merc Diku Mud improvments copyright (C) 1992, 1993 by Michael          *
 *  Chastain, Michael Quan, and Mitchell Tse.                              *
 *                                                                         *
 *  In order to use any part of this Merc Diku Mud, you must comply with   *
 *  both the original Diku license in 'license.doc' as well the Merc       *
 *  license in 'license.txt'.  In particular, you may not remove either of *
 *  these copyright notices.                                               *
 *                                                                         *
 *  Much time and thought has gone into this software and you are          *
 *  benefitting.  We hope that you share your changes too.  What goes      *
 *  around, comes around.                                                  *
 ***************************************************************************/

#include <sys/types.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <time.h>
#include "merc.h"
#include "recycle.h"

BAN_DATA *ban_list;

extern FILE *fpReserve;

#define BAN_FILE "../area/ban.txt"

/*
 * save_bans() - Write permanent bans to disk.
 */
void save_bans(void)
{
	BAN_DATA *pban;
	FILE *fp;
	bool found = FALSE;

	fclose(fpReserve);
	if ((fp = fopen(BAN_FILE, "w")) == NULL) {
		perror(BAN_FILE);
		fpReserve = fopen(NULL_FILE, "r");
		return;
	}

	for (pban = ban_list; pban != NULL; pban = pban->next) {
		if (IS_SET(pban->ban_flags, BAN_PERMANENT)) {
			found = TRUE;
			fprintf(fp, "%-20s %-2d %d\n",
				pban->name, pban->level, pban->ban_flags);
		}
	}

	fclose(fp);
	fpReserve = fopen(NULL_FILE, "r");

	if (!found)
		unlink(BAN_FILE);
}

/*
 * load_bans() - Read bans from disk.
 */
void load_bans(void)
{
	FILE *fp;
	BAN_DATA *ban_last;

	if ((fp = fopen(BAN_FILE, "r")) == NULL)
		return;

	ban_last = NULL;
	for (;;) {
		BAN_DATA *pban;

		if (feof(fp)) {
			fclose(fp);
			return;
		}

		pban = new_ban();

		pban->name      = str_dup(fread_word(fp));
		pban->level     = fread_number(fp);
		pban->ban_flags = fread_flag(fp);
		fread_to_eol(fp);

		if (ban_list == NULL)
			ban_list = pban;
		else
			ban_last->next = pban;
		ban_last = pban;
	}
}

/*
 * check_ban() - Check if a descriptor's host is banned.
 *
 * Called when a new connection is accepted.  If the site is banned,
 * the descriptor is closed.
 */
void check_ban(DESCRIPTOR_DATA *d)
{
	BAN_DATA *pban;
	char host[MAX_STRING_LENGTH];

	if (d == NULL || d->host == NULL)
		return;

	strcpy(host, capitalize(d->host));
	host[0] = LOWER(host[0]);

	for (pban = ban_list; pban != NULL; pban = pban->next) {
		if (!IS_SET(pban->ban_flags, BAN_ALL))
			continue;

		if (IS_SET(pban->ban_flags, BAN_PREFIX)
			&& IS_SET(pban->ban_flags, BAN_SUFFIX)
			&& strstr(host, pban->name) != NULL)
		{
			write_to_buffer(d, "Your site has been banned.\n\r", 0);
			close_socket(d);
			return;
		}

		if (IS_SET(pban->ban_flags, BAN_PREFIX)
			&& !str_suffix(pban->name, host))
		{
			write_to_buffer(d, "Your site has been banned.\n\r", 0);
			close_socket(d);
			return;
		}

		if (IS_SET(pban->ban_flags, BAN_SUFFIX)
			&& !str_prefix(pban->name, host))
		{
			write_to_buffer(d, "Your site has been banned.\n\r", 0);
			close_socket(d);
			return;
		}
	}
}

File: embermud/ban.py
"""Site ban system.

Ports ban.c from the C version.
"""

from __future__ import annotations

import os

from . import game
from .area_loader import fread_flag, fread_number, fread_to_eol, fread_word
from .config import BAN_ALL, BAN_PERMANENT, BAN_PREFIX, BAN_SUFFIX
from .data import BanData
from .log import log_string
from .util import capitalize, str_prefix, str_suffix

import os as _os
BAN_FILE = _os.path.join(
    _os.path.dirname(_os.path.dirname(_os.path.abspath(__file__))),
    "area", "ban.txt"
)


def save_bans() -> None:
    """Write permanent bans to disk."""
    found = False
    try:
        with open(BAN_FILE, "w") as fp:
            for ban in game.ban_list:
                if ban.ban_flags & BAN_PERMANENT:
                    found = True
                    fp.write(f"{ban.name:<20s} {ban.level:<2d} {ban.ban_flags}\n")
    except OSError:
        log_string(f"save_bans: could not open {BAN_FILE}")
        return

    if not found:
        try:
            os.unlink(BAN_FILE)
        except OSError:
            pass


def load_bans() -> None:
    """Read bans from disk."""
    try:
        fp = open(BAN_FILE, "r")
    except OSError:
        return

    try:
        while True:
            line = fp.readline()
            if not line:
                break
            line = line.strip()
            if not line:
                continue
            parts = line.split()
            if len(parts) < 3:
                continue
            ban = BanData(
                name=parts[0],
                level=int(parts[1]),
                ban_flags=int(parts[2]),
            )
            game.ban_list.append(ban)
    finally:
        fp.close()


def check_ban(host: str) -> bool:
    """Check if a host is banned. Returns True if banned."""
    if not host:
        return False

    for ban in game.ban_list:
        if not (ban.ban_flags & BAN_ALL):
            continue

        if (ban.ban_flags & BAN_PREFIX) and (ban.ban_flags & BAN_SUFFIX):
            if ban.name.lower() in host.lower():
                return True

        if ban.ban_flags & BAN_PREFIX:
            if not str_suffix(ban.name, host):
                return True

        if ban.ban_flags & BAN_SUFFIX:
            if not str_prefix(ban.name, host):
                return True

    return False
